File: erp_bot/src/nlu/compound.py
"""Detect and split compound transaction messages."""

from __future__ import annotations

import re

_AMOUNT_COMMA = re.compile(r"\d{1,3}(?:,\d{2,3})+(?:\.\d+)?")
_SPLIT = re.compile(
    r"\s+(?:ra|ani|and|;|\+)\s+"
    r"|"
    r",\s+(?=[a-zA-Z\u0900-\u097F])",
    re.I,
)
_HAS_AMOUNT = re.compile(r"\d|saya|hajar|lakh|panch|das|bees", re.I)
_TX_VERB = re.compile(
    r"\b("
    r"becheko|beche|bechyo|bikri|bikyo|bikne|"
    r"kineko|kinya|kharid|kinyo|"
    r"tiryo|tireko|diye|diyeko|deko|"
    r"kharcha|kharch|expense|bill|"
    r"salary|talab|bhada|bhaada|kiraya|rent|"
    r"jama|aayo|aayeko|liyo|"
    r"payment|paid|sold|purchase|bought|"
    r"advance|deposit|withdraw|"
    r"return|firta|refund"
    r")\b",
    re.I,
)
_LEADING_FILLER = re.compile(r"^(?:aaja|aja|hijo|bholi|yo|aile)\s+", re.I)


def _protect_amount_commas(text: str) -> tuple[str, dict[str, str]]:
    placeholders: dict[str, str] = {}

    def repl(match: re.Match[str]) -> str:
        key = f"__AMT{len(placeholders)}__"
        placeholders[key] = match.group(0)
        return key

    return _AMOUNT_COMMA.sub(repl, text), placeholders


def _restore_amount_commas(text: str, placeholders: dict[str, str]) -> str:
    out = text
    for key, value in placeholders.items():
        out = out.replace(key, value)
    return out


def _part_is_transaction(part: str) -> bool:
    return bool(_HAS_AMOUNT.search(part) and (_TX_VERB.search(part) or len(part) > 14))


def split_compound_transactions(text: str) -> list[str]:
    """
    Split messages like 'aaja bikri 8500, rent 8000 tiryo' into parts.
    Returns original text if not clearly compound.
    """
    t = (text or "").strip()
    if not t or not _HAS_AMOUNT.search(t):
        return [t]

    protected, placeholders = _protect_amount_commas(t)
    parts = [p.strip() for p in _SPLIT.split(protected) if p.strip()]
    parts = [_restore_amount_commas(p, placeholders) for p in parts]
    if len(parts) < 2:
        return [t]

    valid = [p for p in parts if _part_is_transaction(p)]
    if len(valid) < 2:
        return [t]

    # Propagate leading date filler to later parts missing it (aaja bikri 8500, rent 8000)
    shared_prefix = ""
    m = _LEADING_FILLER.match(valid[0])
    if m:
        shared_prefix = m.group(0).strip()
    enriched: list[str] = []
    for part in valid:
        if shared_prefix and not _LEADING_FILLER.match(part) and not _TX_VERB.search(part[:20]):
            enriched.append(f"{shared_prefix} {part}".strip())
        else:
            enriched.append(part)
    return enriched


def is_compound_message(text: str) -> bool:
    return len(split_compound_transactions(text)) > 1


def is_payroll_with_statutory(text: str) -> bool:
    """Salary message mentioning SSF and/or TDS."""
    t = text.lower()
    return bool(re.search(r"\b(salary|talab)\b", t)) and bool(
        re.search(r"\b(ssf|tds|social\s*security|withhold|kataut|katayo)\b", t)
    )


def is_rent_with_tds(text: str) -> bool:
    t = text.lower()
    return bool(re.search(r"\b(rent|bhada|bhaada|kiraya)\b", t)) and bool(
        re.search(r"\b(tiryo|paid|diye|payment|tireko)\b", t)
    )

File: src/lib/ekhata/compound.ts
/** Detect and split compound transaction messages (Python parity). */

const AMOUNT_COMMA = /\d{1,3}(?:,\d{2,3})+(?:\.\d+)?/g;
const SPLIT = /\s+(?:ra|ani|and|;|\+)\s+|,\s+(?=[a-zA-Z\u0900-\u097F])/i;
const HAS_AMOUNT = /\d|saya|hajar|lakh|panch|das|bees/i;
const TX_VERB =
  /\b(becheko|beche|bechyo|bikri|bikyo|bikne|kineko|kinya|kharid|kinyo|tiryo|tireko|diye|diyeko|deko|kharcha|kharch|expense|bill|salary|talab|bhada|bhaada|kiraya|rent|jama|aayo|aayeko|liyo|payment|paid|sold|purchase|bought|advance|deposit|withdraw|return|firta|refund)\b/i;
const LEADING_FILLER = /^(?:aaja|aja|hijo|bholi|yo|aile)\s+/i;

function protectAmountCommas(text: string): {
  protected: string;
  placeholders: Map<string, string>;
} {
  const placeholders = new Map<string, string>();
  const protectedText = text.replace(AMOUNT_COMMA, (match) => {
    const key = `__AMT${placeholders.size}__`;
    placeholders.set(key, match);
    return key;
  });
  return { protected: protectedText, placeholders };
}

function restoreAmountCommas(text: string, placeholders: Map<string, string>): string {
  let out = text;
  for (const [key, value] of placeholders) {
    out = out.replaceAll(key, value);
  }
  return out;
}

function partIsTransaction(part: string): boolean {
  return HAS_AMOUNT.test(part) && (TX_VERB.test(part) || part.length > 14);
}

export function splitCompoundTransactions(text: string): string[] {
  const t = (text || "").trim();
  if (!t || !HAS_AMOUNT.test(t)) return [t];

  const { protected: protectedText, placeholders } = protectAmountCommas(t);
  let parts = protectedText
    .split(SPLIT)
    .map((p) => p.trim())
    .filter(Boolean);
  parts = parts.map((p) => restoreAmountCommas(p, placeholders));
  if (parts.length < 2) return [t];

  const valid = parts.filter(partIsTransaction);
  if (valid.length < 2) return [t];

  let sharedPrefix = "";
  const m = valid[0].match(LEADING_FILLER);
  if (m) sharedPrefix = m[0].trim();

  return valid.map((part) => {
    if (sharedPrefix && !LEADING_FILLER.test(part) && !TX_VERB.test(part.slice(0, 20))) {
      return `${sharedPrefix} ${part}`.trim();
    }
    return part;
  });
}

export function isCompoundMessage(text: string): boolean {
  return splitCompoundTransactions(text).length > 1;
}
